### DISCUSSION TITLE
webots simulation freezes when controller reaches while loop.

### DISCUSSION THREAD

**Discussion by HimGautam** [cyberbotics/webots#4261] [Q&A]
Hi,
I am trying to use webots with NEAT algorithm. I've integrated robot controller with neat algorithm but the controller freezes when it reaches the line ```while robot.step(timestep) != -1:```.  How do I proceed with this? I am also attaching my code and environment for [reference.](https://drive.google.com/file/d/1vNqYXCBfZm9VKPK-VVQyGQmBPHRisUFP/view?usp=sharing)

**Comment by omichel** [comment]
The `robot.step()` function actually blocks until the simulation time progresses. Pressing the "Run" button of Webots should make the `robot.step()` function return.

**Reply by HimGautam** [reply]
Hi @omichel ,
I found the issue, turns out simulation can not run in multiprocessing, please tell me there is a way I can run simulation using multiprocessing, I don't need to render the simulation. There is one more issue that i need help with. When the robot falls over it gets a negative reward, and environment is reset using supervisor thing in my main robot (which fell). But when the environment is resetted remaining script is not compiled, thus the simulation freezes again. How can I deal with this?

**Comment by omichel** [comment]
I would recommend to have the Webots API functions called from the same thread only. Other threads may run in parallel and synchronize with the one dealing with the Webots API.
Your supervisor should recompile the controllers (by invoking `make` in the controller folder) and [restart them](https://cyberbotics.com/doc/reference/supervisor?tab-language=c++#wb_supervisor_node_restart_controller) when resetting the simulation.

**Reply by HimGautam** [reply]
Is there any example that implements this idea. I am a beginner in webots so it is hard for me to follow what you are describing.

**Reply by omichel** [reply]
Not to my knowledge...

**Reply by HimGautam** [reply]
Hi,
I found a solution for the problem. It requires that I run supervisor and then reset the robots physics and translation when the episode is over. But due to some reason the line below do not work on my robot. Can you please help with this?
                   
    supervisor = Supervisor()
    robot_node = supervisor.getFromDef("GhostDog")   <------- (returns None)

**Reply by omichel** [reply]
This means that in your world file, there is no `DEF GhostDog ` node definition. You can check it by opening the world file in a text editor and search for this string.

**Comment by HimGautam** [comment]
Is there anyway third person view of simulation moves along the robot. I mean when my robot moves then how can I make sure my screen always follows the robot?

**Reply by BenjaminDeleze** [reply]
The field `follow` in the node `Viewpoint` will allow you to achieve that.
You can have a look at https://cyberbotics.com/doc/reference/viewpoint for more information.

**Reply by ad-daniel** [reply]
Or just right-click the robot, `follow object` > `tracking shot`/`mounted shot`

**Comment by HimGautam** [comment]
Can you please tell me why I am getting this error?
```/usr/local/bin/webots: line 90:  5289 Segmentation fault      (core dumped) "$webots_home/bin/webots-bin" "$@"```

**Reply by HimGautam** [reply]
Every time I run the simulation it just crashes.

**Reply by omichel** [reply]
It might come from your world file. Did you try to start Webots with a different world file? Alternatively, you may also try the [safe mode](https://cyberbotics.com/doc/guide/starting-webots#safe-mode).

**Reply by HimGautam** [reply]
@omichel, I rechecked all the changes I made in my latest run. Upon some investigation, I found that when I turned on rendering of support polygon, my simulation crashed after few training iterations. I request you to please check if it is a bug or it only happens on my machine.

**Reply by omichel** [reply]
Can you please open an issue about it? Please describe precisely how to reproduce this crash so that we can investigate it efficiently.

**Comment by HimGautam** [comment]
Is there a way to identify and interpret the contact points of already available robots using supervisor node. In simple terms, if the robot falls over, is there a way I can identify the robot has fallen using any supervisor features?

**Reply by ad-daniel** [reply]
It varies on the type of robot of course, but a way of determining it could be to check if the center of mass (https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_node_get_center_of_mass) goes below a certain threshold

**Reply by HimGautam** [reply]
It's a quadruped. I need to detect when robot falls over or collides to the boundary of road or objects. The center of mass method can really help in the balancing task but is there something for collision part?
